Add option to dump statistics every n instructions

MartinNowack opened the pull request:
Use `stats-write-after-instructions` to dump klee statistics after
every n instructions. This is disabled by default
Be aware, `stats-write-interval` is enabled by default.

If both features are enabled, statistics are written every n
instructions and every m seconds to the same file.

andreamattavelli commented:
@MartinNowack Works fine in my experiments.
I don't know what @ccadar thinks, but I would prefer to have `stats-write-after-instructions` and `stats-write-interval` as mutually exclusive. That is, it is not possible to specify both a time and a number of instructions, but only one of the two.

ccadar commented on lib/Core/StatsTracker.cpp:
```diff
@@ -259,6 +267,10 @@ StatsTracker::~StatsTracker() {
 void StatsTracker::done() {
+  // Update statistics for reachability analysis
+  if (updateMinDistToUncovered && (statsFile || OutputIStats))
+      computeReachableUncovered();
```
We need to look at where MinDistToUncovered is written.  It looks like only in istats, so we should guard this accordingly

ccadar commented on lib/Core/StatsTracker.cpp:
```diff
@@ -259,6 +267,10 @@ StatsTracker::~StatsTracker() {
+  if (updateMinDistToUncovered && (statsFile || OutputIStats))
+      computeReachableUncovered();
+
   if (statsFile)
```
Sometimes we ask about statsFile, other times about OutputStats, it would be nice to be consistent

ccadar replied:
Sometimes we ask about statsFile, other times about OutputStats, it would be nice to be consistent

ccadar commented on lib/Core/StatsTracker.cpp:
```diff
@@ -310,6 +322,13 @@ void StatsTracker::stepInstruction(ExecutionState &es) {
+      stats::instructions % StatsWriteAfterInstructions.getValue() == 0) {
+    if (updateMinDistToUncovered)
+      computeReachableUncovered();
+    writeStatsLine();
```
Isn't this function already called at line 275?

ccadar commented:
@andreamattavelli , I agree, we should have a check in there.

MartinNowack commented:
@andreamattavelli @ccadar Ok, convinced, I added the additional check.

@ccadar Good comments, it revealed a bug (unlikely to trigger). Before, if stats were not written to disk, the reachability information was not updated during execution.
Now your comments should be addressed including the bug fixed.

andreamattavelli commented on lib/Core/StatsTracker.cpp:
```diff
@@ -86,18 +86,16 @@ namespace {
-                     cl::desc("Approximate number of seconds between run.branches writes (default: 5.0s)"),
-                     cl::init(5.));
-  */
-
```
Why did you remove this code fragment? I think this is not related with your PR.

andreamattavelli commented:
Beside my comment on the unrelated change in the code, I would squash it in a single commit and I believe that it is ready to be merged.
As always, I would add the new option to the documentation.

andreamattavelli commented on lib/Core/StatsTracker.cpp:
```diff
@@ -96,6 +96,11 @@ namespace {
 
+  cl::opt<unsigned> IStatsWriteAfterInstructions(
+      "istats-write-after-instructions", cl::init(0),
+      cl::desc("Write istats after each n instructions, 0 disable it "
```
0 disables it
or: 0 to disable

andreamattavelli commented on lib/Core/StatsTracker.cpp:
```diff
@@ -86,17 +86,20 @@ namespace {
 
+  cl::opt<unsigned> StatsWriteAfterInstructions(
+      "stats-write-after-instructions", cl::init(0),
+      cl::desc("Write statistics after each n instructions, 0 disable it "
```
0 disables it
or, 0 to disable

andreamattavelli commented on lib/Core/StatsTracker.cpp:
```diff
@@ -86,17 +86,20 @@ namespace {
-  */
+  cl::opt<unsigned> IStatsWriteAfterInstructions(
+      "istats-write-after-instructions", cl::init(0),
+      cl::desc("Write istats after each n instructions, 0 disable it "
```
0 disables it
or, 0 to disable

andreamattavelli commented:
@MartinNowack please fix those couple of very minor comments and squash everything to a single commit.
Please also look at my klee.github.io PR [#37](https://github.com/klee/klee.github.io/pull/37) to check that everything is correct.

After this polishing, we are ready to merge.

andreamattavelli commented:
This PR should fix bug #420.

MartinNowack commented:
@andreamattavelli thanks for the review.
I addressed your comments and merged all the stuff to one commit.

I still kept removing the dead code. In my opinion, it's nothing big worth another PR. Is that fine for you?

andreamattavelli commented:
> I still kept removing the dead code. In my opinion, it's nothing big worth another PR. Is that fine for you?

Fine.

MartinNowack commented:
@andreamattavelli @ccadar Thanks for your reviews!
